src/clocksi_vnode.erl

```diff
@@ -110,6 +112,25 @@ read_data_item(Node, TxId, Key, Type, Updates) ->
 async_read_data_item(Node, TxId, Key, Type) ->
     clocksi_readitem:async_read_data_item(Node, Key, Type, TxId, [], {fsm, self()}).
 
+validate_or_read_data_item(Node, TxId, Key, Type, Token, Updates) ->
+    case clocksi_readitem:validate_or_read_data_item(Node, Key, Type, Token, TxId, []) of
+        {ok, {invalid, Snapshot, _}} ->
+            Updates2 = reverse_and_filter_updates_per_key(Updates, Key),
+            Snapshot2 = clocksi_materializer:materialize_eager(Type, Snapshot, Updates2),
+
+            {ok, {invalid, Snapshot2, ?INVALID_OBJECT_TOKEN}};
+        {ok, valid} ->
+            % When valid, there shouldn't be any update to apply here and the
+            % provided Token is necessarily invalid.
+            ?INVALID_OBJECT_TOKEN = Token,
```

Comment:
See comment above: This check might fail?

Reply:
Removed this one. But the other one is fine I believe.